fix #1914 include UTM format in cache details

pstorch opened the pull request:
Extended the `GeopointFormatter` and `CoordinatesFormatSwitcher` to support UTM coords.

samueltardieu commented on main/src/cgeo/geocaching/location/Deg2UTM.java:
```diff
@@ -0,0 +1,80 @@
+
+    public Deg2UTM(final double lat, final double lon) {
+        zone = (int) Math.floor(lon / 6 + 31);
+        if (lat < -72) {
```
There must be a better way :-)

SammysHP replied:
It should be possible to replace this with an array/string index (divide by 8 ...)

samueltardieu commented on main/src/cgeo/geocaching/location/Deg2UTM.java:
```diff
@@ -0,0 +1,80 @@
+        easting = 0.5 * Math.log((1 + Math.cos(lat * Math.PI / 180) * Math.sin(lon * Math.PI / 180 - (6 * zone - 183) * Math.PI / 180)) / (1 - Math.cos(lat * Math.PI / 180) * Math.sin(lon * Math.PI / 180 - (6 * zone - 183) * Math.PI / 180))) * 0.9996 * 6399593.62 / Math.pow((1 + Math.pow(0.0820944379, 2) * Math.pow(Math.cos(lat * Math.PI / 180), 2)), 0.5) * (1 + Math.pow(0.0820944379, 2) / 2 * Ma ...
+        easting = Math.round(easting * 100) * 0.01;
+        northing = (Math.atan(Math.tan(lat * Math.PI / 180) / Math.cos((lon * Math.PI / 180 - (6 * zone - 183) * Math.PI / 180))) - lat * Math.PI / 180) * 0.9996 * 6399593.625 / Math.sqrt(1 + 0.006739496742 * Math.pow(Math.cos(lat * Math.PI / 180), 2)) * (1 + 0.006739496742 / 2 * Math.pow(0.5 * Math.log((1 + Math.cos(lat * Math.PI / 180) * Math.sin((lon * Math.PI / 180 - (6 * zone - 183) * Math.P ...
+        if (letter < 'M') {
```
Isn't it clearer with `if (lat < 0)`?

samueltardieu commented:
I'm in favor of it, but the conversion code is ugly (I understand that you have taken it from a SO answer). The absence of tests will not help with a safe rewrite.

pstorch commented:
@samueltardieu you are absolutely right. I'm sorry. That was just a quick Hack in the evening with some manual Tests. As I don't understand the algorithm I even can't support it. Let's find something else. I'll close for the time beeing.

Bananeweizen commented:
I suggest to add 2 or 3 of the 10 test cases shown near the bottom of http://www.uwgb.edu/dutchs/UsefulData/UTMFormulas.htm. If those are green, the code can be taken as is, to my mind. That page basically derives the formulas above.

For the huge switch case, I would go with "CDE...WX".charAt(lat/8 + 9) or so.

samueltardieu commented:
We could use all ten of them (computation is fast). However, how would UTM be useful to geocaching? UTM coordinates do not designate a unique point.

Bananeweizen commented:
I vaguely remember someone to say that UTM is the only direct coordinate input system (i.e. not adress input) to his car navigation system.

pstorch commented:
Ok, let's try to improve it. I added the testcases from the page @Bananeweizen mentioned, but I get very different results:

```
lat, lon, easting, northing
70.579277, 45.599420, 522244.040000, 7830584.440000
10.018894, 23.313324, 753580.410000, 1108392.400000
19.479896, 75.662049, 569479.630000, 2154062.080000
21.072465, 29.828684, 793928.830000, 2332777.750000
5.458957, 36.385237, 210253.780000, 604026.410000
70.175454, 22.865350, 570594.250000, 7786520.620000
61.965605, 58.931371, 601263.280000, 6871855.180000
11.116050, 20.901069, 489196.660000, 1228812.960000
32.210543, 60.705849, 283776.690000, 3566081.350000
79.187451, 61.532382, 469267.500000, 8791281.830000
```

The calculated values for easting and northing are very different. But when I compare the results with other converters (e.g. GCC App or http://www.rcn.montana.edu/resources/converter.aspx) it looks like the `Deg2UTM` class is quite right.

pstorch commented:
Maybe the 10 Testcases on the HTML Page can't be used as they are, because of this hint below: 
> Note that all longitudes and eastings are with respect to a central meridian of zero degrees and zero easting.

I've done another test with data from the Excel Sheet, Tab "Test Data", Cells K23 - P62. This looks much better and correspond to other conversion tools, but it seams that they exposed serious bugs in the `Deg2UTM` class. There are three testcases with way off results (!!!).

```
lat, lon; expected easting, northing; result easting, northing; diff easting, northing
-55.980000, -67.289170; expected 606750.000000, 3794825.000000; result 606749.900000, 3794825.240000; diff 0.09999999997671694, -0.24000000022351742
-61.453333, -55.495752; expected 580191.000000, 3185792.000000; result 580191.470000, 3185791.900000; diff -0.4699999999720603, 0.10000000009313226
37.819720, -122.478610; expected 545889.000000, 4185941.000000; result 545888.640000, 4185940.900000; diff 0.35999999998603016, 0.10000000009313226
40.713000, -74.013500; expected 583326.000000, 4507366.000000; result 583325.620000, 4507366.040000; diff 0.3800000000046566, -0.0400000000372529
3.157850, 101.711650; expected 801396.000000, 349434.000000; result 801396.250000, 349433.960000; diff -0.25, 0.03999999997904524
27.988060, 86.925280; expected 492652.000000, 3095881.000000; result 492652.530000, 3095881.950000; diff -0.5300000000279397, -0.9500000001862645
!!! -3.075830, 37.353330; expected 317004.000000, 9659884.000000; result 317003.840000, -340116.040000; diff 0.15999999997438863, 1.000000004E7
25.197140, 55.274110; expected 326103.000000, 2787892.000000; result 326102.980000, 2787892.320000; diff 0.02000000001862645, -0.31999999983236194
41.008550, 28.979940; expected 666499.000000, 4541594.000000; result 666499.130000, 4541594.270000; diff -0.1300000000046566, -0.27000000048428774
41.902220, 12.453330; expected 288761.000000, 4642057.000000; result 288760.640000, 4642056.310000; diff 0.35999999998603016, 0.6899999994784594
48.858220, 2.294500; expected 448252.000000, 5411935.000000; result 448251.820000, 5411934.890000; diff 0.17999999999301508, 0.11000000033527613
71.172500, 25.784440; expected 456220.000000, 7897075.000000; result 456219.560000, 7897075.400000; diff 0.4400000000023283, -0.40000000037252903
-34.358060, 18.471940; expected 267496.000000, 6195246.000000; result 267495.650000, 6195245.940000; diff 0.34999999997671694, 0.05999999959021807
!!! 83.627000, -32.664000; expected 504211.000000, 9286900.000000; result 504163.900000, 9286465.650000; diff 47.09999999997672, 434.34999999962747
-55.980000, -67.289170; expected 606750.000000, 3794825.000000; result 606749.900000, 3794825.240000; diff 0.09999999997671694, -0.24000000022351742
-61.453333, -55.495752; expected 580191.000000, 3185792.000000; result 580191.470000, 3185791.900000; diff -0.4699999999720603, 0.10000000009313226
37.819720, -122.478610; expected 545889.000000, 4185941.000000; result 545888.640000, 4185940.900000; diff 0.35999999998603016, 0.10000000009313226
40.713000, -74.013500; expected 583326.000000, 4507366.000000; result 583325.620000, 4507366.040000; diff 0.3800000000046566, -0.0400000000372529
3.157850, 101.711650; expected 801396.000000, 349434.000000; result 801396.250000, 349433.960000; diff -0.25, 0.03999999997904524
27.988060, 86.925280; expected 492652.000000, 3095881.000000; result 492652.530000, 3095881.950000; diff -0.5300000000279397, -0.9500000001862645
!!! -3.075830, 37.353330; expected 317004.000000, 9659884.000000; result 317003.840000, -340116.040000; diff 0.15999999997438863, 1.000000004E7
25.197140, 55.274110; expected 326103.000000, 2787892.000000; result 326102.980000, 2787892.320000; diff 0.02000000001862645, -0.31999999983236194
41.008550, 28.979940; expected 666499.000000, 4541594.000000; result 666499.130000, 4541594.270000; diff -0.1300000000046566, -0.27000000048428774
41.902220, 12.453330; expected 288761.000000, 4642057.000000; result 288760.640000, 4642056.310000; diff 0.35999999998603016, 0.6899999994784594
48.858220, 2.294500; expected 448252.000000, 5411935.000000; result 448251.820000, 5411934.890000; diff 0.17999999999301508, 0.11000000033527613
71.172500, 25.784440; expected 456220.000000, 7897075.000000; result 456219.560000, 7897075.400000; diff 0.4400000000023283, -0.40000000037252903
-34.358060, 18.471940; expected 267496.000000, 6195246.000000; result 267495.650000, 6195245.940000; diff 0.34999999997671694, 0.05999999959021807
-13.163330, -72.545560; expected 766062.000000, 8543503.000000; result 766061.120000, 8543503.770000; diff 0.8800000000046566, -0.7699999995529652
-32.653430, -70.011080; expected 405179.000000, 6386681.000000; result 405179.510000, 6386681.350000; diff -0.5100000000093132, -0.35000000055879354
-43.595750, 170.141040; expected 430668.000000, 5172667.000000; result 430667.660000, 5172666.350000; diff 0.3399999999674037, 0.6499999994412065
-33.858670, 151.214030; expected 334786.000000, 6252080.000000; result 334786.670000, 6252079.350000; diff -0.6699999999837019, 0.6499999994412065
25.033610, 121.565000; expected 355224.000000, 2769437.000000; result 355224.200000, 2769436.600000; diff -0.20000000001164153, 0.39999999990686774
35.358060, 138.731110; expected 293848.000000, 3915114.000000; result 293848.030000, 3915114.500000; diff -0.030000000027939677, -0.5
56.056110, 160.644170; expected 602389.000000, 6213543.000000; result 602389.390000, 6213543.360000; diff -0.39000000001396984, -0.3600000003352761
71.388890, -156.479170; expected 589769.000000, 7922642.000000; result 589768.590000, 7922642.340000; diff 0.4100000000325963, -0.3399999998509884
21.365000, -157.950000; expected 608862.000000, 2362907.000000; result 608862.260000, 2362907.470000; diff -0.2600000000093132, -0.47000000020489097
-27.116670, -109.366670; expected 661897.000000, 6999591.000000; result 661896.480000, 6999590.390000; diff 0.5200000000186265, 0.6099999994039536
-37.111110, -12.288330; expected 740947.000000, 5889360.000000; result 740946.890000, 5889360.070000; diff 0.10999999998603016, -0.07000000029802322
-77.850000, 166.666670; expected 539154.000000, 1357814.000000; result 539154.240000, 1357813.660000; diff -0.23999999999068677, 0.34000000008381903
```

SammysHP commented:
I still recommend a proper conversation library. Not sure if there exists one, though.

Bananeweizen commented:
*sigh* Okay, if we can't get it to work reliable, then I'm also in favor of not adding it, unfortunately.

pstorch commented:
I did some more tests. This time with the IBM sourcecode from here: https://www.ibm.com/developerworks/java/library/j-coordconvert/index.html
I's only one class which offers some more conversions. Its result has only one difference to the testcases from the Excel Sheet and here the testcase looks wrong. Other conversion tools have the same result as the IBM class.
But the IBM class doesn't look like clean code, either. There is no provided maven/gradle package, so we would have to include the source into our project and maintain it. And I'm not sure about the license. On the IBM developerWorks download page it looks rather good, but in the class itself it says `(c) Copyright IBM Corp. 2007`.
I'm searching on...

Bananeweizen commented:
having (and declaring) a copyright and granting a license are two separate things. To me the IBM license on the download page also looks compatible enough with our Apache license.

pstorch commented:
@Bananeweizen copyright vs license makes sense. I'll try to adapt the IBM class a bit to our codebase. Then we can decide.

pstorch commented:
I used the IBM class now and added more tests. The IBM class has better results, but fails at 4 tests with wrong zone letters. I verified the result with several other online converters (mentioned in the test class comment).
I also spent some time cleaning it up, but still way to go (e.g. codacy issues).
Not sure at the moment if it is worth it. A small ready to use and maintained lib (as @SammysHP suggested) would be better I guess.

pstorch commented:
Out of curiosity I went a little bit forward with the IBM derived class.
This UTM coordinate format is a bit strange. I fixed some ZONE Letter issues with North- and Southpole. Added some testcases from geocaching.com and found that even they do it wrong here ;)
The grid has also some Exception (e.g. Norway with Zone 32V, which is wider than the others). The code doesn't handle it correctly. I guess we have some users in Norway who want to have that correct.
Still doubts about it.
At least we have some testcases here to test an external library, if we find one.

pstorch commented:
Just an info: did a quick test with Jcoord from here: http://www.jstott.me.uk/jcoord/
It fails the tests with North- and Southpole Zones. The 1.1 beta version even throws an Exception if a Coordinate falls into that zones. And it's not available via a Maven repository.

pstorch commented:
This library https://github.com/Berico-Technologies/Geo-Coordinate-Conversion-Java is an extract of the NASA World Wind project. There is no artifact in the maven repos, either.
It doesn't provide the latitude zone letters at all and also has problems with coords at the poles.

pstorch commented:
Another lib: https://github.com/Jotschi/geoconvert
Similar to the others: no maven artifact provided, doesn't handle the coords at the poles and this has also problems with the Norway Zone Exception.

samueltardieu commented:
Remind me why UTM would be useful at all? :-)

pstorch commented:
Another old library project: https://sites.google.com/site/ahmdalitaha2/latlonglib
Except the missing maven artifact and the usual pole zone issue the results are ok.

@samueltardieu yes, that's more and more my opinion, too ;) 
I don't know why it is kind of popular in Geocaching (e.g. Mysteries and field puzzles). I've used it myself in one of my caches. 

There are some java libraries around, but most of them old, not well supported and buggy.
I think at least we have to ignore the pole zones (A,B, Y and Z) which are seldom used in Geocaching anyway.

samueltardieu commented:
Have you tried openmap? http://openmap.bbn.com/doc/api/com/bbn/openmap/proj/coords/package-summary.html

pstorch commented:
Tried openmap right now:

- it's the first project which seams not abandoned: good
- Except the polar coodinate zones (always returns Z), it produces good test results: good enough I guess
- it's a full standalone app, it adds 3.7 MB to our libs directory with lots of AWT stuff in it: not so good

The 3.7 MB jar file really hurts from my point of view. What do you think?

Bananeweizen commented:
The 3.7 MB of the file are not really an indication. You need to build a release version with and without that library to see the actual difference. Proguard will strip out everything not needed (as far as it can detect), so the actual difference might be much smaller.

I don't like their license, since it is a custom license. However, it _seems_ to be matching our needs (as far as I can read legal stuff).

pstorch commented:
Ok, the resulting APK size is not so much bigger.

cgeo-debug.apk 
- with openmap.jar: 7.629.103 bytes
- withou openmap.jar: 7.489.010 bytes

Had to add a bunch of proguard `-dontwarn` entries.

There is still one obstacle: I also wanted to implement the parsing of UTM (which was requested by @Lineflyer here https://github.com/cgeo/cgeo/issues/1914#issuecomment-13404482), but this is not possible due to a dependency to `java.awt.geom.Point2D` in the LatLongPoint class of openmap.

Bananeweizen commented:
maybe you can file a bug against the project to remove the dependency. I don't know whether that was a design choice of them or just an oversight.

SammysHP commented on main/src/cgeo/geocaching/location/UTMConverter.java:
```diff
@@ -4,44 +4,42 @@
         @Override
         public String toString() {
-            return longZone + latZone + " " + Math.round(easting) + " "
+            return longZone + latZone + " E " + Math.round(easting) + " N "
```
Usually E and N are omitted.

pstorch replied:
Geocaching.com is showing them. But of course that doesn't mean it should be done that way  ;)

pstorch commented:
I think changing openmap to fit our needs is not feasible. The `java.awt` classes are deeply interwoven in the openmap codebase. It's not an easy task to decouple the calculation logic from the UI logic.
The interims solution I made was to duplicate some methods which don't use the `java.awt` dependent class and substituted it with a simple double array for latitude and longitude. But this doesn't fit in their original design.
One solution could be to extract and adjust the needed classes from openmap. Either into cgeo directly or into its own library project.
While writing this comment another developer commented in the openmap issue (https://github.com/OpenMap-java/openmap/issues/31#issuecomment-236285647) that he/she has implemented an UTM conversion lib and wants to send it on monday. Let's see.

BTW: it's interesting how an initially small task (as I thought) evolves into a big cross project task. Normally I would say the UTM conversion is such a seldom feature, it's not worth the effort. But at the moment I'm a little bit hooked. Can't believe there is no usable java/android library for it.

pstorch commented:
I continue with this one. As adjusting openmap is not feasable I try to extract the necessary classes as source and adapt them for cgeo.
Still work in progress. Code needs to be cleanenup, I expect codacy to go wild on this ;)

pstorch commented:
I think it is ready for another review/test round now.
The resulting build errors are not related to my changes.

The calculations in `UTMPoint` have still room for improvements. I adapted the code from openmap to some extend. Feedback welcome. The tests should cover the UTM conversion pretty good. So we can play around with the math and improve it.

I've added the parsing logic, too. So if the user has an UTM string in the clipboard he/she can use it to create a new Waypoint.

pstorch commented:
retest this please

Bananeweizen commented:
retest this please

pstorch commented:
retest this please

samueltardieu commented on main/src/cgeo/geocaching/location/GeopointParser.java:
```diff
@@ -74,6 +75,14 @@ public static Geopoint parse(@NonNull final String text) {
+        try {
+            final UTMPoint utmPoint = new UTMPoint(text);
+            return utmPoint.toLatLong();
+        } catch (final Exception e) {
```
We tend to call those exceptions `ignore` or `ignored` in new code because some tools get this as a hint that we do not expect to do anything with the exception.

samueltardieu commented on main/src/cgeo/geocaching/location/UTMPoint.java:
```diff
@@ -0,0 +1,367 @@
+     *         <code>null</code> if conversion failed.
+     */
+    public Geopoint toLatLong() {
+        // check the ZoneNummber is valid
```
Typo (two "m").

samueltardieu commented on main/src/cgeo/geocaching/location/UTMPoint.java:
```diff
@@ -0,0 +1,367 @@
+     * North latitudes are positive, South latitudes are negative.
+     *
+     * @return Geopoint containing the lat/long value, or
+     *         <code>null</code> if conversion failed.
```
We support from `null` so much… Wouldn't it make more sense to throw an exception there, and have this function be annotated with `NonNull`?

samueltardieu commented on main/src/cgeo/geocaching/location/UTMPoint.java:
```diff
@@ -0,0 +1,367 @@
+                        * (d * d / 2 - (5 + 3 * t1 + 10 * c1 - 4 * c1 * c1 - 9 * ECC_PRIME_SQUARED) * d * d * d * d / 24 + (61 + 90
+                        * t1 + 298 * c1 + 45 * t1 * t1 - 252 * ECC_PRIME_SQUARED - 3 * c1 * c1)
+                        * d * d * d * d * d * d / 720);
+        lat = radToDeg(lat);
```
Why risk using mutable variables instead of using

``` java
   return new Geopoint(radToDeg(lat), radToDeg(lon));
```

?

(and maybe use `latRad` and `lonRad` as variable names)

samueltardieu commented on main/src/cgeo/geocaching/location/UTMPoint.java:
```diff
@@ -0,0 +1,367 @@
+            utmNorthing += FALSE_NORTHING;
+        }
+
+        return new UTMPoint(zoneNumber, zoneLetter, utmEasting, utmNorthing);
```
If the latest argument was `geopoint.getLatitude() < 0f ? utmNorthing + FALSE_NORTHING : utmNorthing` then `utmNorthing` would not need to be mutable.

samueltardieu commented on main/src/cgeo/geocaching/location/UTMPoint.java:
```diff
@@ -0,0 +1,367 @@
+     * @param rad radians
+     * @return double decimal degrees
+     */
+    private static double radToDeg(final double rad) {
```
Are those proxies really needed?

samueltardieu commented on main/src/cgeo/geocaching/location/UTMPoint.java:
```diff
@@ -0,0 +1,367 @@
+
+        // Make sure the longitude 180.00 is in Zone 60
+        if (lon == 180) {
+            zoneNumber = 60;
```
Why not use early `return` constructs everywhere instead of `zoneNumber =` ?

samueltardieu commented on main/src/cgeo/geocaching/location/UTMPoint.java:
```diff
@@ -0,0 +1,367 @@
+        char letterDesignator = 'Z';
+
+        if ((84 >= lat) && (lat >= 72)) {
+            letterDesignator = 'X';
```
Early `return` should be used there.

pstorch commented:
@samueltardieu good suggestions, thanks for the feedback

samueltardieu commented:
(of course you do not need approval, I'm just playing with the new review system to indicate that I've read the code again)

pstorch commented:
It's a nice feature

samueltardieu commented:
Incidentally, I stumbled accross a mystery cache in Paris with UTM coordinates mid-August, I could have used this feature :-)

Bananeweizen commented:
I only noticed this afterwards: Next time please use AssertJ assertions, not plain Junit assertions. Those are much easier to read and write, and are therefore used in all other tests.
